### PR TITLE
Breezeing over regtest with Breeze.Daemon.exe command-line parameters

### PR DESCRIPTION
**Pull request by maciejzaleski**
Breeze daemon and GUI changes to allow configuration of tumbling protocol when running in **Regtest** network. This functionality is purposely not available in **Testnet** or **Mainnet**.

With the `-noTor` you can tumble without using Tor network. The `-tumblingProtocol` switch allows to change the default communication channel (i.e. _TCPHandler_) to standard HTTP requests handler. The benefit of this change is that the protocol can now be debugged with _WebProxy_ tools.
Without Tor the onion address defaults to the MachineName and it is used only as label (not the actual hostname). The connection is established using address in _IPv4_ field of the registration record.

Other changes include enhancements to the command line parameters which now work in both development mode and the packaged version. This allows to modify ports on which the daemons are run (i.e. API and protocol), datadir, network, use of Tor and tumbling protocol (i.e. Http, Tcp).

### CODE REVIEW

File: Breeze.BreezeServer/Program.cs [hunk 1]
```diff
@@ -84,7 +118,7 @@ public static void Main(string[] args)
 
 			var tumbler = serviceProvider.GetService<ITumblerService>();
 			
-			tumbler.StartTumbler(config, false);
+			tumbler.StartTumbler(config, false, torMandatory: !useTor, tumblerProtocol: tumblerProtocol);
```
Comment (zeptin) [inline Breeze.BreezeServer/Program.cs]: Here too

File: Breeze.TumbleBit.Client.Tests/TumbleBitTests.cs [hunk 2]
```diff
@@ -541,14 +554,16 @@ public void TestWithoutTor()
                         }
                     }
 
-                    Thread.Sleep(20000);
+                    Thread.Sleep(5000);
```
Comment (zeptin) [inline Breeze.TumbleBit.Client.Tests/TumbleBitTests.cs]: Be careful setting this too low, there are delays in the runtime itself (although they are artificially decreased for regtest). It also takes time to perform the state machine updates in the crypto-heavy phases when multiple cycles are running concurrently. I found that even 10 seconds was marginally stable. 20 seconds slows down the test but at least seems to be a workable setting.

File: Breeze.TumbleBit.Client.Tests/TumbleBitTests.cs [hunk 3]
```diff
@@ -838,14 +861,23 @@ public void TestDualClientWithoutTor()
                         }
                     }*/
 
-                    Thread.Sleep(20000);
+                    Thread.Sleep(5000);
```
Comment (zeptin) [inline Breeze.TumbleBit.Client.Tests/TumbleBitTests.cs]: Same here

File: Breeze.BreezeServer/BreezeConfiguration.cs [hunk 4]
```diff
@@ -18,6 +18,8 @@ public class BreezeConfiguration
     {
         public Network TumblerNetwork { get; set; }
 
+	    public bool UseTor { get; set; }
```
[the remark this answers is not in the document]
Comment (maciejzaleski) [reply Breeze.BreezeServer/BreezeConfiguration.cs]: Replaced tabs with spaces

File: Breeze.BreezeServer/BreezeConfiguration.cs [hunk 5]
```diff
@@ -73,7 +76,12 @@ public BreezeConfiguration(string configPath, string datadir = null)
                     TumblerNetwork = Network.TestNet;
                 }
 
-                if (configFile.GetOrDefault<string>("network", "testnet").Equals("regtest"))
+	            if (configFile.GetOrDefault<string>("tor.enabled", "true").Equals("true"))
+	            {
+		            UseTor = true;
```
[the remark this answers is not in the document]
Comment (maciejzaleski) [reply Breeze.BreezeServer/BreezeConfiguration.cs]: I have set the false value explicitly

File: Breeze.BreezeServer/Program.cs [hunk 6]
```diff
@@ -15,6 +15,37 @@ public class Program
...
+			var comparer = new CommandlineArgumentComparer();
+			var isRegTest = args.Contains("regtest", comparer);
+			var isTestNet = args.Contains("testnet", comparer);
+
+			var useTor = !args.Contains("noTor", comparer);
+
+			TumblerProtocolType? tumblerProtocol = null;
+			try
+			{
+				string tumblerProtocolString = args.Where(a => a.StartsWith("-tumblerProtocol=")).Select(a => a.Substring("-tumblerProtocol=".Length).Replace("\"", "")).FirstOrDefault();
```
[the remark this answers is not in the document]
Comment (maciejzaleski) [reply Breeze.BreezeServer/Program.cs]: I wouldn't think we need to document those switches. I have only added them to make testing easier. There is probably no benefit to the end user in changing ports or data directory location.
I will ask on the stand up tomorrow what others think.

File: Breeze.BreezeServer/Program.cs [hunk 7]
```diff
@@ -32,30 +63,33 @@ public static void Main(string[] args)
...
 
             BreezeConfiguration config = new BreezeConfiguration(configPath);
+			if (!useTor)
+				config.UseTor = false;
 
 			logger.LogInformation("{Time} Pre-initialising server to obtain parameters for configuration", DateTime.Now);
 			
 			var preTumblerConfig = serviceProvider.GetService<ITumblerService>();
-			preTumblerConfig.StartTumbler(config, true);
+			preTumblerConfig.StartTumbler(config, true, torMandatory: !useTor, tumblerProtocol: tumblerProtocol);
```
[the remark this answers is not in the document]
Comment (maciejzaleski) [reply Breeze.BreezeServer/Program.cs]: This was a defect. I have tweeted the code to make tor mandatory unless we are running in the testnet

File: Breeze.BreezeServer/Program.cs (hunk 1 above)
Comment (maciejzaleski) [reply Breeze.BreezeServer/Program.cs]: Resolved

File: Breeze.TumbleBit.Client.Tests/TumbleBitTests.cs (hunk 2 above)
Comment (maciejzaleski) [reply Breeze.TumbleBit.Client.Tests/TumbleBitTests.cs]: Restored to original value

File: Breeze.TumbleBit.Client.Tests/TumbleBitTests.cs (hunk 3 above)
Comment (maciejzaleski) [reply Breeze.TumbleBit.Client.Tests/TumbleBitTests.cs]: Restored to original value

File: NTumbleBit/Utils.cs [hunk 8]
```diff
@@ -201,5 +202,16 @@ public static void DeleteRecursivelyWithMagicDust(string destinationDir)
 			}
 			// depending on your use case, consider throwing an exception here
 		}
+
+		public static IPAddress GetInternetConnectedAddress()
```
[the remark this answers is not in the document]
Comment (maciejzaleski) [reply NTumbleBit/Utils.cs]: Agreed

**Pull request merged**